### PR TITLE
HIVE-26059: Eventually clean compactions in "refused" state from compaction history

### PR DESCRIPTION
**Pull request by veghlaci05**
### What changes were proposed in this pull request?
Purge refused compactions from COMPLETED_COMPACTIONS.

### Why are the changes needed?
Compactions in states succeeded, failed, and did not initiate have a retention threshold (example: metastore.compactor.history.retention.succeeded) and are purged from COMPLETED_COMPACTIONS if the number of compactions in this state per partition/unpartitioned table passes the threshold. This behaviour should also be applied to compactions in REFUSED state.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Manually, and through unit tests.

### CODE REVIEW

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java [hunk 1]
```diff
@@ -1092,16 +1092,18 @@ public void updateCompactorState(CompactionInfo ci, long compactionTxnId) throws
   }
 
   private static class RetentionCounters {
-    int didNotInitiateRetention = 0;
-    int failedRetention = 0;
-    int succeededRetention = 0;
+    int didNotInitiateRetention;
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java:1095]: should we mark them as final, to make sure they are initialized?

**Review by deniskuzZ — APPROVED** [review]

**Review by deniskuzZ — CHANGES_REQUESTED** [review]
There is an issue with COMPACTIONS sys table, refused state should be fetched from COMPLETED_COMPACTIONS table

File: metastore/scripts/upgrade/hive/hive-schema-4.0.0-alpha-1.hive.sql [hunk 2]
```diff
@@ -1200,7 +1200,7 @@ SELECT
   CC_PARTITION,
   CASE WHEN CC_TYPE = 'i' THEN 'minor' WHEN CC_TYPE = 'a' THEN 'major' ELSE 'UNKNOWN' END,
   CASE WHEN CC_STATE = 'f' THEN 'failed' WHEN CC_STATE = 's' THEN 'succeeded'
-    WHEN CC_STATE = 'a' THEN 'did not initiate' ELSE 'UNKNOWN' END,
+    WHEN CC_STATE = 'a' THEN 'did not initiate' WHEN 'c' THEN 'refused' ELSE 'UNKNOWN' END,
```
Comment (deniskuzZ) [inline metastore/scripts/upgrade/hive/hive-schema-4.0.0-alpha-1.hive.sql:1203]: ````WHEN CC_STATE = 'c'````

File: metastore/scripts/upgrade/hive/upgrade-3.1.0-to-4.0.0-alpha-1.hive.sql [hunk 3]
```diff
@@ -318,7 +318,7 @@ SELECT
   CC_PARTITION,
   CASE WHEN CC_TYPE = 'i' THEN 'minor' WHEN CC_TYPE = 'a' THEN 'major' ELSE 'UNKNOWN' END,
   CASE WHEN CC_STATE = 'f' THEN 'failed' WHEN CC_STATE = 's' THEN 'succeeded'
-    WHEN CC_STATE = 'a' THEN 'did not initiate' ELSE 'UNKNOWN' END,
+    WHEN CC_STATE = 'a' THEN 'did not initiate'  WHEN 'c' THEN 'refused' ELSE 'UNKNOWN' END,
```
Comment (deniskuzZ) [inline metastore/scripts/upgrade/hive/upgrade-3.1.0-to-4.0.0-alpha-1.hive.sql:321]: ````WHEN CC_STATE = 's'````

Comment (deniskuzZ) [reply metastore/scripts/upgrade/hive/upgrade-3.1.0-to-4.0.0-alpha-1.hive.sql:321]: also you didn't remove `refused` state from COMPACTION_QUEUE

File: metastore/scripts/upgrade/hive/hive-schema-4.0.0-alpha-1.hive.sql (hunk 2 above)
Comment (deniskuzZ) [reply metastore/scripts/upgrade/hive/hive-schema-4.0.0-alpha-1.hive.sql:1203]: also you didn't remove `refused` state from COMPACTION_QUEUE

**Review by deniskuzZ — APPROVED** [review]
pending tests

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/conf/MetastoreConf.java [hunk 4]
```diff
@@ -417,6 +417,10 @@ public static ConfVars getMetaConf(String name) {
         "hive.compactor.history.retention.succeeded", 3,
         new RangeValidator(0, 100), "Determines how many successful compaction records will be " +
         "retained in compaction history for a given table/partition."),
+    COMPACTOR_HISTORY_RETENTION_REFUSED("metastore.compactor.history.retention.refused",
+        "hive.compactor.history.retention.refused", 3,
+        new RangeValidator(0, 100), "Determines how many refused compaction records will be " +
+        "retained in compaction history for a given table/partition."),
     COMPACTOR_HISTORY_RETENTION_TIMEOUT("metastore.compactor.history.retention.timeout",
             "hive.compactor.history.retention.timeout", 7, TimeUnit.DAYS,
             "Determines how long failed and not initiated compaction records will be " +
```
Comment (klcopp) [inline standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/conf/MetastoreConf.java:426]: "Refused" should be listed here as well

**Review by klcopp — APPROVED** [review]
LGTM, one suggestion to edit a comment

**Review by klcopp — APPROVED** [review]

**Pull request merged**